feat(components): [form-item]: add `label-position` prop

xing403 opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

可对单个 form-item 进行label 定位
![image](https://github.com/element-plus/element-plus/assets/74746021/91a39ac1-69de-4b88-a5a9-7800804b9365)
![image](https://github.com/element-plus/element-plus/assets/74746021/a1299d08-c647-40c1-a8db-2a72e31db5c8)
![image](https://github.com/element-plus/element-plus/assets/74746021/6b29bbc1-ae18-46d4-af47-be6df11853f9)

xing403 committed:
feat(components): [form-item]: add `label-position` prop

xing403 commented:
这是使用的例子

```vue
<template>
  <el-form-item label="form label-position" label-position="top">
    <el-radio-group v-model="labelPosition" size="large">
      <el-radio-button v-for="position in options" :key="position" :label="position" :value="position" />
    </el-radio-group>
  </el-form-item>
  <el-form :model="form" label-width="100px" :label-position="labelPosition" style="width: 500px">
    <el-form-item label="top" label-position="top">
      <el-input v-model="form.name" />
    </el-form-item>
    <el-form-item label="default">
      <el-col :span="11">
        <el-date-picker v-model="form.date1" type="date" placeholder="Pick a date" style="width: 100%" />
      </el-col>
      <el-col :span="2" class="text-center">
        <span class="text-gray-500">-</span>
      </el-col>
      <el-col :span="11">
        <el-time-picker v-model="form.date2" placeholder="Pick a time" style="width: 100%" />
      </el-col>
    </el-form-item>
    <el-form-item label="left" label-position="left">
      <el-radio-group v-model="form.resource">
        <el-radio value="Sponsor">Sponsor</el-radio>
        <el-radio value="Venue">Venue</el-radio>
      </el-radio-group>
    </el-form-item>
    <el-form-item label="right" label-position="right">
      <el-input v-model="form.desc" type="textarea" />
    </el-form-item>
  </el-form>
</template>

<script setup lang="ts">
import { reactive, ref } from 'vue'
let labelPosition = ref<'left' | 'right' | 'top'>('left')
const options: Array<'left' | 'right' | 'top'> = ['left', 'right', 'top']
const form = reactive({
  name: '',
  date1: '',
  date2: '',
  resource: '',
  desc: '',
})
</script>
```

xxholly32 commented:
https://github.com/element-plus/element-plus/discussions/16927

之前我也有提过相关讨论~

btea commented on packages/components/form/src/form-item.vue:
```diff
@@ -101,7 +101,8 @@ let initialValue: any = undefined
 
 const labelStyle = computed<CSSProperties>(() => {
-  if (formContext?.labelPosition === 'top') {
+  const labelPosition = props?.labelPosition || formContext?.labelPosition
```
`props?.labelPosition` optional chaining should be removed.

btea commented on packages/components/form/src/form-item.vue:
```diff
@@ -135,7 +137,11 @@ const formItemClasses = computed(() => [
+  {
+    [ns.m('feedback')]: formContext?.statusIcon,
+    [ns.m(`label-${props?.labelPosition || formContext?.labelPosition}`)]:
+      props.labelPosition || formContext?.labelPosition,
```
`props.labelPosition || formContext?.labelPosition` is used in multiple places and can be extracted into a computed value.

btea commented on packages/components/form/src/form-item.ts:
```diff
@@ -27,6 +27,14 @@ export const formItemProps = buildProps({
+  /**
+   * @description Position of label. If set to `'left'` or `'right'`, `label-width` prop is also required. The default is extend from `form label-position`.
+   */
+  labelPosition: {
```
New properties should be added simultaneously in the documentation.

xing403 committed:
docs(components): form-item label-position docs

xing403 committed:
test(components): form-item

closed form-item label-position test

xing403 commented:
> #16927
> 
> 之前我也有提过相关讨论~

我在开发中也遇到了这个需求，觉得这种场景也挺常见的

btea commented on docs/en-US/component/form.md:
```diff
@@ -121,6 +121,16 @@ form/accessibility
 
 :::
 
+## Custom Form Item labelPosition
```
Please add a version tag.
```suggestion
## Custom Form Item labelPosition ^(2.7.6)
```

btea commented on docs/en-US/component/form.md:
```diff
@@ -175,6 +185,7 @@ form/accessibility
 | --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ------- |
 | prop            | A key of `model`. It could be a path of the property (e.g `a.b.0` or `['a', 'b', '0']`). In the use of `validate` and `resetFields` method, the attribute is required. | ^[string] / ^[string&#91;&#93;]                     | —       |
 | label           | Label text.                                                                                                                                                            | ^[string]                                           | —       |
+| label-position  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
```suggestion
| label-position ^(2.7.6)  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```

xing403 committed:
Update docs/en-US/component/form.md

Co-authored-by: btea <2356281422@qq.com>

xing403 committed:
Update docs/en-US/component/form.md

Co-authored-by: btea <2356281422@qq.com>

btea commented on docs/en-US/component/form.md:
```diff
@@ -175,6 +185,7 @@ form/accessibility
 | --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ------- |
 | prop            | A key of `model`. It could be a path of the property (e.g `a.b.0` or `['a', 'b', '0']`). In the use of `validate` and `resetFields` method, the attribute is required. | ^[string] / ^[string&#91;&#93;]                     | —       |
 | label           | Label text.                                                                                                                                                            | ^[string]                                           | —       |
+| label-position ^(2.7.6)  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
```suggestion
| label-position ^(2.7.6)  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`.                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
The document should be formatted to align with the other fields.

xing403 committed:
Update docs/en-US/component/form.md

Co-authored-by: btea <2356281422@qq.com>

btea reviewed (approved):
LGTM. Thanks!

FrontEndDog reviewed (approved):
Thanks for your contribution！

kooriookami commented:
Could you add a complex demo, e.g. [demo](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbjxkaXY+XG48c3Bhbj5Gb3JtIEFsaWduPC9zcGFuPlxuICA8ZWwtcmFkaW8tZ3JvdXAgdi1tb2RlbD1cImxhYmVsUG9zaXRpb25cIiBhcmlhLWxhYmVsPVwibGFiZWwgcG9zaXRpb25cIj5cbiAgICA8ZWwtcmFkaW8tYnV0dG9uIHZhbHVlPVwibGVmdFwiPkxlZnQ8L2VsLXJhZGlvLWJ1dHRvbj5cbiAgICA8ZWwtcmFkaW8tYnV0dG9uIHZhbHVlPVwicmlnaHRcIj5SaWdodDwvZWwtcmFkaW8tYnV0dG9uPlxuICAgIDxlbC1yYWRpby1idXR0b24gdmFsdWU9XCJ0b3BcIj5Ub3A8L2VsLXJhZGlvLWJ1dHRvbj5cbiAgPC9lbC1yYWRpby1ncm91cD5cbjwvZGl2PlxuPGRpdj5cbjxzcGFuPkZpcnN0IEZvcm0gSXRlbSBBbGlnbjwvc3Bhbj5cbiAgPGVsLXJhZGlvLWdyb3VwIHYtbW9kZWw9XCJpdGVtTGFiZWxQb3NpdGlvblwiIGFyaWEtbGFiZWw9XCJsYWJlbCBwb3NpdGlvblwiPlxuICAgIDxlbC1yYWRpby1idXR0b24gdmFsdWU9XCJsZWZ0XCI+TGVmdDwvZWwtcmFkaW8tYnV0dG9uPlxuICAgIDxlbC1yYWRpby1idXR0b24gdmFsdWU9XCJyaWdodFwiPlJpZ2h0PC9lbC1yYWRpby1idXR0b24+XG4gICAgPGVsLXJhZGlvLWJ1dHRvbiB2YWx1ZT1cInRvcFwiPlRvcDwvZWwtcmFkaW8tYnV0dG9uPlxuICA8L2VsLXJhZGlvLWdyb3VwPlxuPC9kaXY+XG4gIDxkaXYgc3R5bGU9XCJtYXJnaW46IDIwcHhcIiAvPlxuICA8ZWwtZm9ybVxuICAgIDpsYWJlbC1wb3NpdGlvbj1cImxhYmVsUG9zaXRpb25cIlxuICAgIGxhYmVsLXdpZHRoPVwiMTUwcHhcIlxuICAgIDptb2RlbD1cImZvcm1MYWJlbEFsaWduXCJcbiAgICBzdHlsZT1cIm1heC13aWR0aDogNjAwcHhcIlxuICA+XG4gICAgPGVsLWZvcm0taXRlbSBsYWJlbD1cIk5hbWVcIiA6bGFiZWwtcG9zaXRpb249XCJpdGVtTGFiZWxQb3NpdGlvblwiPlxuICAgICAgPGVsLWlucHV0IHYtbW9kZWw9XCJmb3JtTGFiZWxBbGlnbi5uYW1lXCIgLz5cbiAgICA8L2VsLWZvcm0taXRlbT5cbiAgICA8ZWwtZm9ybS1pdGVtIGxhYmVsPVwiQWN0aXZpdHkgem9uZVwiPlxuICAgICAgPGVsLWlucHV0IHYtbW9kZWw9XCJmb3JtTGFiZWxBbGlnbi5yZWdpb25cIiAvPlxuICAgIDwvZWwtZm9ybS1pdGVtPlxuICAgIDxlbC1mb3JtLWl0ZW0gbGFiZWw9XCJBY3Rpdml0eSBmb3JtXCI+XG4gICAgICA8ZWwtaW5wdXQgdi1tb2RlbD1cImZvcm1MYWJlbEFsaWduLnR5cGVcIiAvPlxuICAgIDwvZWwtZm9ybS1pdGVtPlxuICA8L2VsLWZvcm0+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVhY3RpdmUsIHJlZiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB0eXBlIHsgRm9ybVByb3BzIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5jb25zdCBsYWJlbFBvc2l0aW9uID0gcmVmPEZvcm1Qcm9wc1snbGFiZWxQb3NpdGlvbiddPigncmlnaHQnKVxuY29uc3QgaXRlbUxhYmVsUG9zaXRpb24gPSByZWY8Rm9ybVByb3BzWydsYWJlbFBvc2l0aW9uJ10+KCdyaWdodCcpXG5cbmNvbnN0IGZvcm1MYWJlbEFsaWduID0gcmVhY3RpdmUoe1xuICBuYW1lOiAnJyxcbiAgcmVnaW9uOiAnJyxcbiAgdHlwZTogJycsXG59KVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctMTcxMTEtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiXG4gIH1cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnsic3R5bGVTb3VyY2UiOiJodHRwczovL3ByZXZpZXctMTcxMTEtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5jc3MifX0=)

May be it can be more prettier.

kooriookami commented on docs/en-US/component/form.md:
```diff
@@ -121,6 +121,16 @@ form/accessibility
 
+## Custom Form Item labelPosition ^(2.7.6)
+
+The label of the component can be positioned using the `label-position` property in el-form-item. The default extend `el-form`
```
```suggestion
You can set the position of `el-form-item` separately using `label-position`. If the value is not set, the `label-position` of `el-form` is used.
```

kooriookami commented:
![image](https://github.com/element-plus/element-plus/assets/38392315/9a00054a-d125-4ab2-8fbe-101cb1eb7e8c)

May be we can merge the two demos in one demo.

xing403 commented:
> ![image](https://private-user-images.githubusercontent.com/38392315/341704618-9a00054a-d125-4ab2-8fbe-101cb1eb7e8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NjAzNDYsIm5iZiI6MTcxODk2MDA0NiwicGF0aCI6Ii8zODM5MjMxNS8zNDE3MDQ2MTgtOWEwMDA1NGEtZDEyNS00YWIyLThmYmUtMTAxY2IxZWI3ZThjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDA4NTQwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0ZWIxNTBiY2VjYWE3NDU3OTJlYTFhMGJhM2Y5YWVlNjRkMmY4YzE5NDJmOWQ3ODdlYTI0MWQyNzU5YzU3YWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VMFWqfa8jUjYjB0JaRfJpXH18iDM-cnI2PigStH81dc)
> 
> May be we can merge the two demos in one demo.

好的好的

xing403 committed:
fix(components): form-item line-height is overridden by the form style
fix(components): form-item label-position style is invalid when label-width is auto

xing403 committed:
docs(components):  update form and form-item alignment examples

kooriookami commented on docs/en-US/component/form.md:
```diff
@@ -168,33 +158,33 @@ form/form-item-labelPosition
+| ----------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ------- |
+| prop                    | A key of `model`. It could be a path of the property (e.g `a.b.0` or `['a', 'b', '0']`). In the use of `validate` and `resetFields` method, the attribute is required. | ^[string] / ^[string&#91;&#93;]                     | —       |
+| label                   | Label text.                                                                                                                                                            | ^[string]                                           | —       |
+| label-position ^(2.7.6) | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`.                                | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
```suggestion
| label-position ^(2.7.7) | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`.                                | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```

kooriookami commented on docs/en-US/component/form.md:
```diff
@@ -45,7 +45,7 @@ form/inline-form
 ## Alignment
 
-Depending on your design, there are several different ways to align your label element.
+Depending on your design, there are several different ways to align your label element. You can set the position of `el-form-item` separately using `label-position`. If the value is not set, the `label-position` of `el-form` is used.
```
```suggestion
Depending on your design, there are several different ways to align your label element. You can set the position of `el-form-item` separately using `label-position` ^(2.7.7). If the value is not set, the `label-position` of `el-form` is used.
```

xing403 committed:
Update docs/en-US/component/form.md

Co-authored-by: kooriookami <38392315+kooriookami@users.noreply.github.com>

xing403 committed:
Update docs/en-US/component/form.md

Co-authored-by: kooriookami <38392315+kooriookami@users.noreply.github.com>

kooriookami reviewed (approved):
LGTM
